diff --git a/pywal/export.py b/pywal/export.py
index 47ea4d2..e2e3f5b 100644
--- a/pywal/export.py
+++ b/pywal/export.py
@@ -53,8 +53,8 @@ def template(colors, input_file, output_file=None):
                     replace_str += '.' + fname
                     new_color = function
 
-            if isinstance(new_color, util.Color):
-                new_color = new_color.strip
+            #if isinstance(new_color, util.Color):
+            #    new_color = new_color.strip
             # If the color was changed, replace with a unique identifier.
             if new_color is not colors[cname]:
                 new_color = str(new_color)
